node/validator_store.go

```diff
@@ -0,0 +1,540 @@
+		bundleData := batchDatum.BundleBytes
+
+		bundleKeyBuilder, err := s.levelDB.GetKeyBuilder(BundleTableName)
+		if err != nil {
+			return nil, 0, fmt.Errorf("failed to get key builder for bundles: %v", err)
+		}
+		bundleKey := bundleKeyBuilder.Key(bundleKeyBytes)
+
+		keys = append(keys, bundleKey)
+		dbBatch.PutWithTTL(bundleKey, bundleData, s.ttl)
+		size += uint64(len(bundleData) + len(bundleKey.Raw()))
+
+	}
+
+	if err := dbBatch.Apply(); err != nil {
+		return nil, 0, fmt.Errorf("failed to apply batch: %v", err)
+	}
+
+	return keys, size, nil
+}
+
+func (s *validatorStore) storeBatchLittDB(batchData []*BundleToStore) (uint64, error) {
+	var size uint64
+
+	writeCompleteChan := make(chan error, len(batchData))
+	for _, batchDatum := range batchData {
+		bundleKeyBytes := batchDatum.BundleKey
+		bundleData := batchDatum.BundleBytes
+
+		go func() {
+			// Grab a lock on the hash of the blob. This protects against duplicate writes of the same blob.
+			lockIndex := uint64(util.HashKey(bundleKeyBytes[:], s.duplicateRequestSalt))
+			s.duplicateRequestLock.Lock(lockIndex)
+			defer s.duplicateRequestLock.Unlock(lockIndex)
+
+			exists, err := s.chunkTable.Exists(bundleKeyBytes[:])
+			if err != nil {
+				writeCompleteChan <- fmt.Errorf("failed to check existence: %v", err)
+				return
+			}
+
+			if exists {
+				// Data is already present, no need to write it again.
+				writeCompleteChan <- nil
+				return
+			}
+
+			err = s.chunkTable.Put(bundleKeyBytes, bundleData)
+			if err != nil {
+				writeCompleteChan <- fmt.Errorf("failed to put data: %v", err)
+			}
+
+			writeCompleteChan <- nil
+		}()
+
+		size += uint64(len(bundleKeyBytes) + len(bundleData))
+	}
+
+	for i := 0; i < len(batchData); i++ {
+		err := <-writeCompleteChan
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	err := s.chunkTable.Flush()
+	if err != nil {
+		return 0, fmt.Errorf("failed to flush chunk table: %v", err)
+	}
+
+	return size, nil
+}
+
+func (s *validatorStore) DeleteKeys(keys []kvstore.Key) error {
+	if s.littDB != nil {
+		return fmt.Errorf("littDB does not support deletion")
+	}
+
+	dbBatch := s.levelDB.NewTTLBatch()
+	for _, key := range keys {
+		dbBatch.Delete(key)
+	}
+	return dbBatch.Apply()
+}
+
+func (s *validatorStore) GetBundleData(bundleKey []byte) ([]byte, error) {
+	if s.littDB == nil {
+		// We haven't thrown the switch for littDB yet. Just look in levelDB.
+		return s.getChunksLevelDB(bundleKey)
+	}
+
+	// Regardless of migration status, always check littDB first.
+	data, ok, err := s.getChunksLittDB(bundleKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get chunks: %v", err)
+	}
+
+	if !ok {
+		// The data wasn't found in littDB.
+
+		if s.levelDB == nil {
+			// There is no data in levelDB.
+			return nil, fmt.Errorf("failed to get chunks: not found")
+		}
+
+		s.migrationLock.RLock()
+		defer s.migrationLock.RUnlock()
+
+		if s.levelDB == nil {
+			// The migration completed while we were waiting for the lock.
```

Comment:
would it make sense to reattempt a `get` from littdb in this case?

Reply:
Nope. This condition can only trigger in the following circumstances:

- we've looked for the data, and it isn't in littDB
- the data may not be in levelDB either, in which case we do no harm saying "we can't find it"
- the data may be in levelDB
    - the data is about to expire
    - the data is some of the last data in levelDB to expire
    - when the user asks for the data, it has not yet expired
    - before the data is returned to the user, the last data expires in levelDB, and we delete the entire database

Migration is not a true "migration", in that we aren't moving data from levelDB into littDB. We are actually just waiting for all of the levelDB data to exceed its TTL, after which we delete the database in its entirety.